prow/crier/controller.go

```diff
@@ -240,37 +240,37 @@ func (c *Controller) processNextItem() bool {
 	}
 
 	logrus.WithField("prowjob", keyRaw).Infof("Will report state : %s", pj.Status.State)
-
-	if err := c.reporter.Report(pj); err != nil {
+	pjs, err := c.reporter.Report(pj)
```

Comment:
Gerrit doesn't have individual job context or statuses, multiple jobs share a common label, so we need to aggregate job reports to report all the jobs that share a label.

Reply:
Previously we report multiple jobs but only has one job set the PrevReportState which is causing trouble for determine the re-triggering reporting logic.

(Gerrit doesn't have context so we have to rely on the `PrevReportState` state on the prowjob)

Reply:
I think a more robust system would be able to determine if a state was or was not reported. Having a `Report()` signature that takes a prowjob and returns a slice of prowjobs seems ... bad. Or, can you report then use a separate call to get other prowjobs that also need to get marked as reported?

Reply:
Or, if a new version of the job was triggered, shouldn't the old one get marked aborted?

Reply:
IMHO the current design tries to let gerrit reporter only handle gerrit side of logics, and let crier handles prowjob updates.

For example, the update call will return an updated object, similarly report call should return actual reported objects, which is consistent to me.

Use a separate call to get prowjobs reported in this cycle seems rather harder to implement & understand.

Reply:
But you're changing the interface for all reporters used in `crier` -- asking `crier` to report on one PJ and getting a list of _other_ PJ back seems totally unreasonable for anything other than this use case

Reply:
It won't change functionality for other reporters though, why not accommodate the change and maybe one day we want to do aggregated report for github as well :-)

Reply:
Because the interface itself is ridiculous and leaking one specific impl into an interface for all reporters is bad? Why is it hard to determine if prow job A should not get reported because prow job B already did? And do the filter before calling out to the report lib?

Reply:
We can add a new method to the interface which does the filtering, and make it no-op for other reporters, if that makes more sense to you.

Reply:
Why not filter to the set of jobs that do need reporting first, since that is what the Gerrit controller should know how to do, then only call report on the list that needs reporting?

Reply:
Who will call gerrit adapter to get that list? Crier doesn't have knowledge about gerrit though, and crier need to know that before call into the report method for each reporter.

Reply:
Hrm. Why doesn't this affect us on GitHub with aborted ProwJobs?

Reply:
hummmmmm - abort state is set by plank separately in https://github.com/kubernetes/test-infra/blob/master/prow/plank/controller.go#L200, we only report if state changes within `syncProwJobs` afaik.

Reply:
But what about when `crier` is handling the reporting? Will it have the same issue as `gerrit` here?

Reply:
We can skip aborted jobs in [`ShouldReport`](https://github.com/kubernetes/test-infra/blob/master/prow/github/reporter/reporter.go#L54)

Reply:
That seems appropriate -- is that the only case where we want to filter out?

Reply:
That's not the only case, although that is one of the case we want to handle.

The root issue is, gerrit doesn't allow you to edit comments. Say we have presubmit {A, B, C}, without aggregate, we will report whenever A or B or C finishes (imagine we have 10s of presubmits, the bot will spam 10 times under a change, which is the original complaint). 

Instead we need to wait until all three jobs to finish, so whenever a job finishes, it will check the status of other jobs, the last finished job will gather all the jobs and do the report. That's why somewhere (pre-filter before Report, or during the Report call), we need to gather that list of need to reported prowjobs, and set the prev report state properly.

Reply:
Seems like we should not try to have one report interface if we want different functionality, and the fact that we're editing the single shared interface to have reporter-specific functionality breaks our abstraction here. 

/shrug

I think it will be clearer -- if we agree  that we want to approach reporting differently based on the output type -- to just have different implementations. Reading one "generic" workflow with caveats that are only ever triggered per reporter type is more complex than reading a workflow where every branch can actually be meaningful

Reply:
I want to keep a unified interface which will make new reporter easier to implement (like the pubsub one ended up pretty trivial).

Does go allow anonymous return types like a struct{}? Then we can let crier figure out if the returned value is a single prowjob, or a list of prowjobs?

Reply:
The point of a strongly typed interface is to abstract common functionality, though. Side-stepping the type system isn't generally a good idea. Since we're already diverging the behavior behind the reporter, I'm not sure there is a large amount of merit to having an interface (or at least, to using it for `gerrit`)